swarm/pss/notify/notify.go

```diff
@@ -0,0 +1,276 @@
+package notify
+
+import (
+	"bytes"
+	"crypto/ecdsa"
+	"fmt"
+	"sync"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/p2p"
+	"github.com/ethereum/go-ethereum/rlp"
+	"github.com/ethereum/go-ethereum/swarm/pss"
+)
+
+const (
+	// sent from requester to updater to request start of notifications
+	MsgCodeStart = iota
+
+	// sent from updater to requester, contains a notification plus a new symkey to replace the old
+	MsgCodeNotifyWithKey
+
+	// sent from updater to requester, contains a notification
+	MsgCodeNotify
+
+	// sent from requester to updater to request stop of notifications (currently unused)
+	MsgCodeStop
+	MsgCodeMax
+)
+
+const (
+	DefaultAddressLength = 1
+	symKeyLength         = 32 // this should be gotten from source
+)
+
+var (
+	// control topic is used before symmetric key issuance completes
+	controlTopic = pss.Topic{0x00, 0x00, 0x00, 0x01}
+)
+
+// when code is MsgCodeStart, Payload is address
+// when code is MsgCodeNotifyWithKey, Payload is notification | symkey
+// when code is MsgCodeNotify, Payload is notification
+// when code is MsgCodeStop, Payload is address
+type Msg struct {
+	Code    byte
+	Name    []byte
+	Payload []byte
+}
+
+func NewMsg(code byte, name string, payload []byte) *Msg {
+	return &Msg{
+		Code:    code,
+		Name:    []byte(name),
+		Payload: payload,
+	}
+}
+
+func (self *Msg) GetName() string {
+	return string(self.Name)
+}
+
+// a notifier has one sendmux entry for each address space it sends messages to
+type sendBin struct {
+	address  pss.PssAddress
+	symKeyId string
+	count    int
+}
+
+// represents a single notification service
+// only subscription address bins that match the address of a notification client have entries. The threshold sets the amount of bytes each address bin uses.
+// every notification has a topic used for pss transfer of symmetrically encrypted notifications
+// contentFunc is the callback to get initial update data from the notifications service provider
+type notifier struct {
+	bins        []*sendBin
+	topic       pss.Topic
+	threshold   int
+	contentFunc func(string) ([]byte, error)
+}
+
+// Controller is the interface to control, add and remove notification services
+type Controller struct {
+	pss       *pss.Pss
+	notifiers map[string]*notifier
+	handlers  map[string]func(string, []byte) error
+	mu        sync.Mutex
+}
+
+// NewController creates a new Controller object
+func NewController(ps *pss.Pss) *Controller {
+	ctrl := &Controller{
+		pss:       ps,
+		notifiers: make(map[string]*notifier),
+		handlers:  make(map[string]func(string, []byte) error),
+	}
+	ctrl.pss.Register(&controlTopic, ctrl.Handler)
+	return ctrl
+}
+
+// IsActive is used to check if a notification service exists for a specified id string
+// Returns true if exists, false if not
+func (self *Controller) IsActive(name string) bool {
```

Comment:
Nitpick, but in Go, it is generally accepted not to use `self` and `this`, because it doesn't convey any meaning. In this case `c` would be better and just reading a longer method, seeing `c` one knows it refers to the `Controller`.

https://github.com/golang/go/wiki/CodeReviewComments#receiver-names

Reply:
Oh wow. There's a lot of sinning against this in the code. MAybe a good first issue to fix it?

Reply:
I think it is also against the linter (I see that Doug is currently fixing this in another PR). So probably just worth search/replace this now to `c *Controller`